compiler/rustc_typeck/src/check/pat.rs

```diff
@@ -663,6 +663,46 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 ast::Mutability::Not => "",
             };
 
+            let mut_var_suggestion = 'block: {
+                if !matches!(mutbl, ast::Mutability::Mut) {
+                    break 'block None;
+                }
+
+                let ident_kind = match binding_parent {
+                    hir::Node::Param(_) => Some("parameter"),
+                    hir::Node::Local(_) => Some("variable"),
+                    hir::Node::Arm(_) => Some("binding"),
+
+                    // Provide diagnostics only if the parent pattern is struct-like,
+                    // i.e. where `mut binding` makes sense
+                    hir::Node::Pat(Pat { kind, .. }) => match kind {
+                        PatKind::Struct(..)
+                        | PatKind::TupleStruct(..)
+                        | PatKind::Or(..)
+                        | PatKind::Tuple(..)
+                        | PatKind::Slice(..) => Some("binding"),
+
+                        PatKind::Wild
+                        | PatKind::Binding(..)
+                        | PatKind::Path(..)
+                        | PatKind::Box(..)
+                        | PatKind::Ref(..)
+                        | PatKind::Lit(..)
+                        | PatKind::Range(..) => None,
+                    },
+
+                    // Don't provide suggestions in other cases
+                    _ => None,
+                };
+
+                ident_kind.map(|thing| (
+                    pat.span,
+                    format!("to declare a mutable {thing} use `mut variable_name`"),
```

Comment:
I prefer `to declare a mutable {thing} use:` -- the `mut variable_name` part is redundant with the suggestion that follows direectly after